### ISSUE TITLE
Release Agda 2.6.0

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#3391]
I think we should start working seriously on getting Agda 2.6.0 out, hopefully before Christmas. Hence I hereby declare a **feature freeze** of the `master` branch until the release (new features should go to `experimental` instead).

I will start to go over all the issues currently labeled with `2.6.0` and move the ones that are not critical to `2.6.1`.

**Comment by UlfNorell** [comment]
> new features should go to `experimental` instead

It's called `future`, but otherwise yes. If you add something to `future`, make sure to bring it up to date with current `master` first. It's currently sitting somewhere close to `2.5.4`.

**Comment by jespercockx** [comment]
Oh, I remember having a discussion about how we should name this branch, but apparently I misremembered the outcome :S

**Comment by asr** [comment]
Before the release, the closed issues with milestone `2.6.0` will be added to the `CHANGELOG` except the issues with [these](https://github.com/agda/agda/blob/a9d4392fa13efb63e886a3bbf8df59fbdde74d15/HACKING.md#closing-issues) labels. Please add a missing label if necessary when closing issues.

**Comment by jespercockx** [comment]
Phew, after a first pass the number of open issues with 2.6.0 milestone went down from >250 to 133 that I think we should still look at.

Open issues by assignee:
- no assignee: 76
- @jespercockx 19
- @andreasabel 13
- @UlfNorell 9
- @Saizan 6
- @asr 3
- @gallais 2
- @fredrikNordvallForsberg 1

Still lots of work to do, but I'm confident we can get it done!

**Comment by jespercockx** [comment]
Since we just missed our initial deadline (which was perhaps a tad optimistic), I propose the following adjustment to the plan:

- Have an initial release candidate ready before Christmas
- Release Agda 2.6.0 before the 13th of January (start of POPL)

**Comment by jespercockx** [comment]
I just did another pass over all open issues and bumped several that seemed unlikely to be fixed before the release. This leaves us with just 37 issues remaining before we can push out the release. Go Agda team!

**Comment by gallais** [comment]
Are we still on the feature freeze? I could see merging small things such as
#3456, #3485, and #3548.

I would hold off on #3458 and #3533 because they're more involved and even
though they seem OK, more testing before an inclusion in 2.6.1 won't do any
harm.

**Comment by UlfNorell** [comment]
Those seem perfectly fine. The feature freeze is more of a recommendation than a hard rule. Use your own judgement to decide.

**Comment by jespercockx** [comment]
We're now very close to a release, there are only 5 open issues left. Thanks to everyone for the hard work! 

Remaining issues:
- [x] https://github.com/agda/agda/issues/3539
- [x] https://github.com/agda/agda/issues/3537
- [ ] https://github.com/agda/agda/issues/3536
- [x] https://github.com/agda/agda/issues/1556
- [x] https://github.com/agda/agda/issues/2774

It would be great if we could get a release candidate now so we can release before the Agda meeting next month. Maybe we could already have a RC with listing these as 'known issues'? @asr do you have time to make the RC?

**Comment by asr** [comment]
> Maybe we could already have a RC with listing these as 'known issues'?

I'll list these known issues in the RC. I'm waiting for green on Travis before starting to make the RC.

**Comment by Saizan** [comment]
I suppose all the issues with label `cubical` should also be `not-in-changelog` for this release? Or what is the policy?

**Comment by asr** [comment]
> I suppose all the issues with label `cubical` should also be `not-in-changelog` for this release?

Right. Please add the `not-in-changelog` label.

**Comment by asr** [comment]
As preparation for the RC, I *manually* added into CHANGELOG the [list of all closed issues](https://github.com/agda/agda/blob/master/CHANGELOG.md#list-of-all-closed-issues) in 2.6.0 using the `closed-issues-for-milestone` program. If there is an issue which shouldn't be in list, please changes its labels following [this](https://github.com/agda/agda/blob/master/HACKING.md#closing-issues) information.

**Comment by asr** [comment]
As preparation for the RC, I added into the CHANGELOG the [list of all closed issues](https://github.com/agda/agda/blob/e61495c7b49c86a2d20a95835fce673bd1dc9705/CHANGELOG.md) in 2.6.0 (using the `closed-issues-for-milestone` program). If there is an issue which should/shouldn't be in the CHANGELOG, please change its labels following [this](https://github.com/agda/agda/blob/463be33e6b41fb4291d6bef1e23606c0398a2f33/HACKING.md) information.

After changing issue's labels you can:

1. manually fix the CHANGELOG or

2. run the `closed-issues-for-milestone` program by

```
$ cd src/release-tools/closed-issues-for-milestone/
$ cabal install
$ export GITHUBTOKEN=your-personal-access-token
$ closed-issues-for-milestone 2.6.0
```

**Comment by jespercockx** [comment]
Should we include PR's in the list of fixed 'issues'? For example #3411 and #3524 are currently in the list. I would tend to not include them, or possibly include them as a separate list.

**Comment by asr** [comment]
I don't think so. We are not including *what* closed an issue but the *closed* issues. It is unfortunate that GitHub uses the same numbering for issues and PRs. The above examples were included because they have `milestone=2.6.0`. We usually don't add milestones to the PRs.

**Comment by asr** [comment]
Travis is green (well, it should be green in a couple of hours), so I'm starting the procudure for releasing the RC. I'm freezing the RC in this version of master:

```
$ agda --version
Agda version 2.6.0-33826ed
```

**Comment by gallais** [comment]
I think the fact that `--guardedness --safe` is not equivalent to `--safe --guardedness`
is too confusing. Given that we have a fix (#3567), I'm hoping to be able to sneak it in
before the release.

**Comment by asr** [comment]
Since some PRs are fixing not reported issues, we should add they to the list of closed issues in the CHANGELOG. Please add a milestone to the PR if you want it be listed in the CHANGELOG.

**Comment by asr** [comment]
I uploaded to Hackage the package candidate [Agda-2.5.4.2.20190217](https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate) and I tested its installation on Ubuntu with the following versions of GHC:

- 8.6.3
- 8.4.4
- 8.2.2
- 8.0.2
- 7.10.3

Could someone test the installation on macOS and Windows using the following instructions:

```bash
$ cabal install https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate/Agda-2.5.4.2.20190217.tar.gz
```

or

```bash
$ cabal get https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate/Agda-2.5.4.2.20190217.tar.gz
$ cd Agda-2.5.4.2.20190217
$ cabal install
```

**Comment by asr** [comment]
There were some changes in `master` which I want to include in the RC. I'm waiting the merge of #3571 for preparing a new RC.

**Comment by andreasabel** [comment]
I suppose all issues with label:cubical should not appear in the change log.

**Comment by asr** [comment]
Right, see https://github.com/agda/agda/issues/3391#issuecomment-463398808.

**Comment by martinescardo** [comment]
It also works with 8.0.2, for the record.

> I uploaded to Hackage the package candidate [Agda-2.5.4.2.20190217](https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate) and I tested its installation on Ubuntu with the following versions of GHC:
> 
>     * 8.6.3
> 
>     * 8.4.4
> 
>     * 8.2.2
> 
>     * 8.0.2
> 
>     * 7.10.3
> 
> 
> Could someone test the installation on macOS and Windows using the following instructions:
> 
> ```shell
> $ cabal install https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate/Agda-2.5.4.2.20190217.tar.gz
> ```
> 
> or
> 
> ```shell
> $ cabal get https://hackage.haskell.org/package/Agda-2.5.4.2.20190217/candidate/Agda-2.5.4.2.20190217.tar.gz
> $ cd Agda-2.5.4.2.20190217
> $ cabal install
> ```

**Comment by jespercockx** [comment]
I'm currently testing the installation on Windows.

**Comment by jespercockx** [comment]
Installation works and I managed to typecheck the `experimental` branch of the standard library. This is on Windows 10 using GHC 8.4.3 and Cabal 2.2.0.0 (which was what I happened to have installed).

**Comment by asr** [comment]
I'm starting the procedure for creating a new RC.

Edited: I'm freezing this RC in 44f5fb6.

**Comment by gallais** [comment]
We should probably wait for #3622 to be fixed.

**Comment by asr** [comment]
Ok. I'll wait.

**Comment by asr** [comment]
I uploaded to Hackage the new package candidate [Agda-2.5.4.2.20190310](https://hackage.haskell.org/package/Agda-2.5.4.2.20190310/candidate) and I tested its installation on Ubuntu with the following versions of GHC:

- 8.6.4
- 8.4.4
- 8.2.2
- 8.0.2
- 7.10.3

Could someone test the installation on macOS and Windows using the following instructions:

```bash
$ cabal install https://hackage.haskell.org/package/Agda-2.5.4.2.20190310/candidate/Agda-2.5.4.2.20190310.tar.gz
```

or

```bash
$ cabal get https://hackage.haskell.org/package/Agda-2.5.4.2.20190310/candidate/Agda-2.5.4.2.20190310.tar.gz
$ cd Agda-2.5.4.2.20190310
$ cabal install
```

**Comment by UlfNorell** [comment]
- [x] MacOS X 10.14, ghc-8.6.3
- [x] Windows 10, ghc-8.4.3

**Comment by jespercockx** [comment]
All set for the release!

**Comment by jespercockx** [comment]
We should go over the release notes to rearrange them and add some highlights to the beginning; currently they are quite unorganized.

**Comment by asr** [comment]
Following the documentation of GHC (see, for example, [Release notes for version 8.6.4](https://downloads.haskell.org/~ghc/8.6.4/docs/html/users_guide/8.6.4-notes.html)), a better place for the highlights could be the user manual.

**Comment by jespercockx** [comment]
I did some reordering and cleaning up of the release notes in https://github.com/agda/agda/commit/c833661d3a813a68a3a42e31a7ca1faf04ad2eba, let me know what you think.

**Comment by MatthewDaggitt** [comment]
Presumably when v2.6.0 officially releases we'll want the standard library submodule to point towards to the v1.0 release of the standard library? Currently this is almost ready, subject only to a few last PRs being merged in. 

@asr do you know when you are planning on releasing? I'll try to sync up timings wise.

**Comment by jespercockx** [comment]
If we do release v1.0 of the standard library together with 2.6.0 then that should definitely also be mentioned in the highlights of the release.

**Comment by MatthewDaggitt** [comment]
> If we do release v1.0 of the standard library together with 2.6.0

We kind of need to because I'm pretty sure v0.17 of the library doesn't work with v2.6.0! So if we don't get the timings right, then Agda would be temporarily without a standard library and civilisation might come crashing down :fearful:

**Comment by asr** [comment]
> I did some reordering and cleaning up of the release notes in https://github.com/agda/agda/commit/c833661d3a813a68a3a42e31a7ca1faf04ad2eba, let me know what you think.

Great! I'll mention the highlights when I announce the release.

**Comment by asr** [comment]
> Presumably when v2.6.0 officially releases we'll want the standard library submodule to point towards to the v1.0 release of the standard library?

Yes!

> @asr do you know when you are planning on releasing?

The RC was announced five days ago. I suggest to wait at least one week but at most two weeks starting from the date when the RC was announced. @MatthewDaggitt, is this time window  enough for you?

**Comment by MatthewDaggitt** [comment]
> The RC was announced five days ago. I suggest to wait at least one week but at most two weeks starting from the date when the RC was announced. @MatthewDaggitt, is this time window enough for you?

Yes. Let's say it'll be ready to go by first thing Monday.

**Comment by asr** [comment]
Good! Please let me know when you get green on Travis after updating the submodule commit to standard library 1.0.

**Comment by MatthewDaggitt** [comment]
Okay, I've fast forwarded the submodule to the right commit, and fixed the tests so they all now pass locally. I've pushed the changes to `release-2.6.0`. Waiting for travis to go green.

**Comment by asr** [comment]
I need the changes in the `master` branch no in the `release-2.6.0` branch.

**Comment by MatthewDaggitt** [comment]
Ah sorry. I always get that wrong 😔. I'm afraid I'm away from my computer
for the next 4 hours. Is it possible for you to cherrypick the last two
commits across? If not I'll fix it when I get home.

**Comment by asr** [comment]
> Is it possible for you to cherrypick the last two commits across?

Yes, I'll do it.

**Comment by asr** [comment]
I cherry-picked (and squashed) the two commits on `master` and I reverted them on `release-2.6.0`.

**Comment by jmchapman** [comment]
I tested the new support for .lagda.md in RC Agda-2.5.4.2.20190310. The ?s don't turn into proper goals. (unless I am doing something stupid).

https://github.com/agda/agda/issues/2837

**Comment by martinescardo** [comment]
I tested it a while ago and it did work for me. Did you run `agda-mode compile` after installing Agda?

**Comment by jmchapman** [comment]
I didn't. I have now. It doesn't seem to have any effect.

**Comment by martinescardo** [comment]
I have Agda version 2.6.0-d6b522c and the following

````
```
f : Set → Set
f X = ?
```
````
does create the hole with a `.lagda.md` file. I will try the pre-release version to see.

**Comment by martinescardo** [comment]
You are right, @jmchapman. With the pre-release version, the "?" remains and the hole is not created.

**Comment by martinescardo** [comment]
It does work with the very latest commit, which is b239cb3.

**Comment by jmchapman** [comment]
@martinescardo thank you for checking!

I pulled the current head `Agda 2.6.0-b71beb8` and can confirm that ?s get correctly turned into goals with this version on my machine.

**Comment by asr** [comment]
I'm planning to merge `master` in `release-2.6.0`, so the new support for `.lagda.md` files will be included in 2.6.0.

**Comment by martinescardo** [comment]
I am going to be teaching at the Midlands Graduate School in about two weeks, and my lecture notes use Agda 2.6.0. Do you think it will be released by then? Or should I write instructions for the students to use the development version or the pre-release? Thanks.

**Comment by asr** [comment]
Is the current master (4d1f6f4063c4cd7c0aea52d567376da79a52f9c7) enough for your course? If so, I can release a new release candidate.

**Comment by martinescardo** [comment]
Thanks. I've just tested this version just in case, and it works fine with both the lecture notes and my personal Agda development, so that this as the new release candidate will be excellent.

**Comment by asr** [comment]
I'm preparing the second release candidate from 4eb2d7d.

**Comment by asr** [comment]
I uploaded to Hackage the second release candidate [Agda-2.5.4.2.20190330](https://hackage.haskell.org/package/Agda-2.5.4.2.20190330/candidate) and I tested its installation on Ubuntu with the following versions of GHC:

- 8.6.4
- 8.4.4
- 8.2.2
- 8.0.2
- 7.10.3

Could someone test the installation on macOS and Windows using the following instructions:

```bash
$ cabal install https://hackage.haskell.org/package/Agda-2.5.4.2.20190330/candidate/Agda-2.5.4.2.20190330.tar.gz
```

or

```bash
$ cabal get https://hackage.haskell.org/package/Agda-2.5.4.2.20190330/candidate/Agda-2.5.4.2.20190330.tar.gz
$ cd Agda-2.5.4.2.20190330
$ cabal install
```

**Comment by jespercockx** [comment]
I managed to install Agda-2.5.4.2.20190330 under Windows 10 with GHC 8.4.3.

**Comment by UlfNorell** [comment]
Installs fine on Mac OS 10.14 and GHC 8.6.3.

**Comment by asr** [comment]
@jespercockx added a *necessary* section of 'highlights' to the `CHANGELOG.md`. I think we should include in this section all the language changes from closed issues with label `language-change`. What do you think?

**Comment by UlfNorell** [comment]
None of those 4 issues look like they deserve a spot in the highlights section. It would be one thing if these were the main things people had to worry about when migrating from 2.5.4 to 2.6.0, but I'm pretty sure that's not the case.

**Comment by MatthewDaggitt** [comment]
Fast-forwarded the standard library again (on `master` this time!) to include a few more bug fixes before v1.0. The commit 92823a9 needs cherry-picking across to release.

**Comment by asr** [comment]
Thanks! I'm planning to merge `master` into `release-2.6.0` when preparing the release.

**Comment by asr** [comment]
All issues with milestone 2.6.0 are closed. Great! I'm waiting for green in Travis to make the release based on `master` from ba50d355f3407c64d948d60042b3e0a31b2b18c6.

**Comment by ice1000** [comment]
FYI CI is already green.

**Comment by andreasabel** [comment]
Fix of #3689 needs to be included in the release (but _not_ in the changelog, as it was a regression caused by new feature #3533).

**Comment by asr** [comment]
Noted.

**Comment by asr** [comment]
I uploaded to Hackage the package candidate [Agda-2.6.0](https://hackage.haskell.org/package/Agda-2.6.0/candidate) and I tested its installation on Ubuntu with GHC 8.6.4, 8.4.4, 8.2.2, 8.0.2 and 7.10.3.

Could someone test the installation on macOS and Windows using the following instructions:

```bash
$ cabal install https://hackage.haskell.org/package/Agda-2.6.0/candidate/Agda-2.6.0.tar.gz
```

or

```bash
$ cabal get https://hackage.haskell.org/package/Agda-2.6.0/candidate/Agda-2.6.0.tar.gz
$ cd Agda-2.6.0
$ cabal install
```

**Comment by asr** [comment]
Please note I'll be without access to any computer for six days starting this Saturday. If I haven't  information about the tests on macOS and Windows soon, I'll not be able to make the release in the coming days.

**Comment by UlfNorell** [comment]
- [x] Windows 10, ghc-8.4.1
- [x] MacOS 10.14, ghc-8.6.4

**Comment by asr** [comment]
Agda 2.6.0 was released. Good job everyone!

**Comment by martinescardo** [comment]
Great! Just in time for the Midlands Graduate School, which starts on Sunday.
